tensorboard/plugins/mesh_visualizer/mesh_plugin.py

```diff
@@ -0,0 +1,263 @@
+    """
+    # Retrieve the multiplexer from the context and store a reference to it.
+    self._multiplexer = context.multiplexer
+    self._tag_to_instance_tags = collections.defaultdict(list)
+    self._instance_tag_to_tag = dict()
+    self._instance_tag_to_metadata = dict()
+    self.prepare_metadata()
+
+  def prepare_metadata(self):
+    """Processes all tags and caches metadata for each."""
+    if self._tag_to_instance_tags:
+      return
+    # This is a dictionary mapping from run to (tag to string content).
+    # To be clear, the values of the dictionary are dictionaries.
+    all_runs = self._multiplexer.PluginRunToTagToContent(MeshPlugin.plugin_name)
+
+    # tagToContent is itself a dictionary mapping tag name to string
+    # SummaryMetadata.plugin_data.content. Retrieve the keys of that dictionary
+    # to obtain a list of tags associated with each run. For each tag, estimate
+    # the number of samples.
+    self._tag_to_instance_tags = collections.defaultdict(list)
+    self._instance_tag_to_metadata = dict()
+    for _, tag_to_content in six.iteritems(all_runs):
+      for tag, content in six.iteritems(tag_to_content):
+        meta = metadata.parse_plugin_metadata(content)
+        self._instance_tag_to_metadata[tag] = meta
+        # Remember instance_name (instance_tag) for future reference.
+        self._tag_to_instance_tags[meta.name].append(tag)
+        self._instance_tag_to_tag[tag] = meta.name
+
+  @wrappers.Request.application
+  def _serve_tags(self, request):
+    """A route (HTTP handler) that returns a response with tags.
+
+    Args:
+      request: The werkzeug.Request object.
+
+    Returns:
+      A response that contains a JSON object. The keys of the object
+      are all the runs. Each run is mapped to a (potentially empty)
+      list of all tags that are relevant to this plugin.
+    """
+    # This is a dictionary mapping from run to (tag to string content).
+    # To be clear, the values of the dictionary are dictionaries.
+    all_runs = self._multiplexer.PluginRunToTagToContent(
+        MeshPlugin.plugin_name)
+
+    # Make sure we populate tags mapping structures.
+    self.prepare_metadata()
+
+    # tagToContent is itself a dictionary mapping tag name to string
+    # SummaryMetadata.plugin_data.content. Retrieve the keys of that dictionary
+    # to obtain a list of tags associated with each run. For each tag estimate
+    # number of samples.
+    response = dict()
+    for run, tag_to_content in six.iteritems(all_runs):
+      response[run] = dict()
+      for instance_tag, _ in six.iteritems(tag_to_content):
+        # Make sure we only operate on user-defined tags here.
+        tag = self._instance_tag_to_tag[instance_tag]
+        meta = self._instance_tag_to_metadata[instance_tag]
+        # Shape should be at least BxNx3 where B represents the batch dimensions
+        # and N - the number of points, each with x,y,z coordinates.
+        assert len(meta.shape) >= 3
```

Comment:
Assertions are for internal consistency checks, not input validation. It
is perfectly possible for this code to reach a state where the assertion
fails. In that case, you should return a 4xx response, not raise an
`AssertionError`.

Reply:
I'll move assert to summary, raise proper exception there and will also make it more strict - require only tensors of shape BxNx3 since I haven't tested other options there.